File: Game-1-modular/systems/geography/region_generator.py
"""Region generator — subdivides nations into geographic regions.

Regions are the large-scale geographic identity of an area.
Biomes derive FROM regions, not the other way around.

Optimized for 512x512 worlds using noise-perturbed Voronoi
(organic boundaries without expensive separate deformation pass).
"""

from __future__ import annotations

from typing import Dict, List, Set, Tuple

from systems.geography.config import GeographicConfig
from systems.geography.models import (
    NationData,
    RegionData,
    RegionIdentity,
)
from systems.geography.noise import (
    find_components,
    hash_2d,
    hash_2d_int,
    voronoi_subdivide,
)


_ALL_IDENTITIES = list(RegionIdentity)


def _determine_region_count(
    nation_id: int,
    nation_chunk_count: int,
    seed: int,
    config: GeographicConfig,
) -> int:
    """Determine how many regions a nation should have."""
    min_r = config.region.min_per_nation
    max_r = config.region.max_per_nation
    range_r = max_r - min_r

    size_factor = min(1.0, nation_chunk_count / 60000.0)
    base = min_r + size_factor * range_r

    variance = hash_2d(nation_id, 0, seed + 33333) * 2.0 - 1.0
    count = int(base + variance * 1.5)

    return max(min_r, min(max_r, count))


def _assign_identity(
    region_idx: int,
    nation_id: int,
    seed: int,
) -> RegionIdentity:
    """Assign a geographic identity to a region."""
    identity_idx = hash_2d_int(region_idx, nation_id, seed + 44444, len(_ALL_IDENTITIES))
    return _ALL_IDENTITIES[identity_idx]


def _fix_contiguity(
    regions: List[Set[Tuple[int, int]]],
) -> List[Set[Tuple[int, int]]]:
    """Ensure each region is contiguous. Merge fragments into neighbors."""
    result = list(regions)

    for i in range(len(result)):
        components = find_components(result[i])
        if len(components) <= 1:
            continue

        components.sort(key=len, reverse=True)
        main = components[0]
        result[i] = main

        for fragment in components[1:]:
            merge_target = _find_adjacent_region(fragment, result, i)
            if merge_target >= 0:
                result[merge_target] |= fragment
            else:
                result[i] |= fragment

    return result


def _find_adjacent_region(
    region: Set[Tuple[int, int]],
    all_regions: List[Set[Tuple[int, int]]],
    exclude_idx: int,
) -> int:
    """Find the index of an adjacent region to merge into."""
    neighbor_counts: Dict[int, int] = {}
    for cx, cy in region:
        for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            neighbor = (cx + dx, cy + dy)
            for idx, other_region in enumerate(all_regions):
                if idx != exclude_idx and neighbor in other_region:
                    neighbor_counts[idx] = neighbor_counts.get(idx, 0) + 1
                    break

    if not neighbor_counts:
        return -1
    return max(neighbor_counts, key=neighbor_counts.get)


def _validate_region_areas(
    regions: List[Set[Tuple[int, int]]],
    nation_chunk_count: int,
    config: GeographicConfig,
) -> List[Set[Tuple[int, int]]]:
    """Merge regions that are too small into neighbors."""
    min_pct = config.region.min_area_pct
    min_area = int(nation_chunk_count * min_pct)

    result = list(regions)
    changed = True
    max_iterations = 10

    while changed and max_iterations > 0:
        changed = False
        max_iterations -= 1
        i = 0
        while i < len(result):
            if len(result[i]) < min_area and len(result) > 1:
                merge_target = _find_adjacent_region(result[i], result, i)
                if merge_target >= 0:
                    result[merge_target] |= result[i]
                    result.pop(i)
                    changed = True
                    continue
            i += 1

    return result


def _compute_bounds(chunks: Set[Tuple[int, int]]) -> Tuple[int, int, int, int]:
    """Compute bounding box (min_x, min_y, max_x, max_y)."""
    if not chunks:
        return (0, 0, 0, 0)
    xs = [c[0] for c in chunks]
    ys = [c[1] for c in chunks]
    return (min(xs), min(ys), max(xs), max(ys))


# ══════════════════════════════════════════════════════════════════
# PUBLIC API
# ══════════════════════════════════════════════════════════════════

def generate_regions(
    nation_map: Dict[Tuple[int, int], int],
    nation_metadata: Dict[int, NationData],
    seed: int,
    config: GeographicConfig,
) -> Tuple[Dict[Tuple[int, int], int], Dict[int, RegionData]]:
    """Generate regions within all nations.

    Uses noise-perturbed Voronoi for organic boundaries without
    expensive separate deformation pass.
    """
    nation_territories: Dict[int, Set[Tuple[int, int]]] = {}
    for pos, nid in nation_map.items():
        nation_territories.setdefault(nid, set()).add(pos)

    region_map: Dict[Tuple[int, int], int] = {}
    region_metadata: Dict[int, RegionData] = {}
    next_region_id = 0

    for nation_id, territory in sorted(nation_territories.items()):
        nation_data = nation_metadata.get(nation_id)
        if not nation_data or not territory:
            continue

        region_seed = seed + nation_id * 1000 + 100000

        # Determine region count
        region_count = _determine_region_count(
            nation_id, len(territory), region_seed, config,
        )

        # Voronoi with noise perturbation for organic boundaries
        raw_regions = voronoi_subdivide(
            territory, region_count, region_seed,
            noise_amplitude=config.region.deform_amplitude,
            noise_frequency=config.region.deform_frequency,
        )

        # Fix contiguity
        fixed_regions = _fix_contiguity(raw_regions)

        # Validate area constraints
        validated_regions = _validate_region_areas(
            fixed_regions, len(territory), config,
        )

        # Assign identities and build metadata
        for i, region_chunks in enumerate(validated_regions):
            if not region_chunks:
                continue

            rid = next_region_id
            next_region_id += 1

            identity = _assign_identity(i, nation_id, region_seed)

            region_metadata[rid] = RegionData(
                region_id=rid,
                name="",
                nation_id=nation_id,
                identity=identity,
                chunk_count=len(region_chunks),
                bounds=_compute_bounds(region_chunks),
            )

            nation_data.region_ids.append(rid)

            for pos in region_chunks:
                region_map[pos] = rid

    return region_map, region_metadata

File: Game-1-Godot/src/Game1.Core/World/Geography/RegionGenerator.cs
namespace Game1.Core.World.Geography;

/// <summary>
/// Port of systems/geography/region_generator.py (post determinism patch):
/// noise-perturbed Voronoi subdivision of each nation into 3-8 regions,
/// contiguity repair, small-region merging, identity assignment by
/// list-index hash. Mutates nation metadata region_ids like Python.
/// </summary>
public static class RegionGenerator
{
    // _determine_region_count — seed here is region_seed
    public static int DetermineRegionCount(int nationId, int nationChunkCount,
                                           long seed, GeographicConfig config)
    {
        var minR = config.Region.MinPerNation;
        var maxR = config.Region.MaxPerNation;
        var rangeR = maxR - minR;
        var sizeFactor = Math.Min(1.0, nationChunkCount / 60000.0);
        var baseCount = minR + sizeFactor * rangeR;
        var variance = GeoNoise.Hash2D(nationId, 0, seed + 33333) * 2.0 - 1.0;
        var count = (int)(baseCount + variance * 1.5);   // int() truncation
        return Math.Max(minR, Math.Min(maxR, count));
    }

    // _assign_identity — index into RegionIdentity declaration order
    public static string AssignIdentity(int regionIdx, int nationId, long seed)
    {
        var identityIdx = GeoNoise.Hash2DInt(regionIdx, nationId, seed + 44444,
                                             GeoTables.RegionIdentities.Length);
        return GeoTables.RegionIdentities[identityIdx];
    }

    // _find_adjacent_region (patched: sorted region iteration) — most
    // adjacency EDGES wins; Python max() first-max-in-insertion-order ties
    public static int FindAdjacentRegion(HashSet<(int X, int Y)> region,
                                         List<HashSet<(int X, int Y)>> allRegions,
                                         int excludeIdx)
    {
        var neighborCounts = new Dictionary<int, int>();
        var neighborOrder = new List<int>();
        foreach (var (cx, cy) in region.OrderBy(p => p))   // sorted(region)
        {
            foreach (var (dx, dy) in new[] { (-1, 0), (1, 0), (0, -1), (0, 1) })
            {
                var neighbor = (cx + dx, cy + dy);
                for (var idx = 0; idx < allRegions.Count; idx++)
                {
                    if (idx != excludeIdx && allRegions[idx].Contains(neighbor))
                    {
                        if (!neighborCounts.ContainsKey(idx))
                        {
                            neighborCounts[idx] = 0;
                            neighborOrder.Add(idx);
                        }
                        neighborCounts[idx] += 1;
                        break;
                    }
                }
            }
        }

        if (neighborCounts.Count == 0) return -1;

        var best = neighborOrder[0];
        var bestCount = neighborCounts[best];
        foreach (var key in neighborOrder)
        {
            if (neighborCounts[key] > bestCount)
            {
                bestCount = neighborCounts[key];
                best = key;
            }
        }
        return best;
    }

    // _fix_contiguity — shares set objects like Python's shallow list copy
    public static List<HashSet<(int X, int Y)>> FixContiguity(
        List<HashSet<(int X, int Y)>> regions)
    {
        var result = new List<HashSet<(int X, int Y)>>(regions);
        for (var i = 0; i < result.Count; i++)
        {
            var components = GeoNoise.FindComponents(result[i]);
            if (components.Count <= 1) continue;

            components = components.OrderByDescending(c => c.Count).ToList();  // stable
            result[i] = components[0];   // rebind to main

            foreach (var fragment in components.Skip(1))
            {
                var mergeTarget = FindAdjacentRegion(fragment, result, i);
                if (mergeTarget >= 0)
                    result[mergeTarget].UnionWith(fragment);
                else
                    result[i].UnionWith(fragment);
            }
        }
        return result;
    }

    // _validate_region_areas — merge tiny regions, ≤10 sweeps
    public static List<HashSet<(int X, int Y)>> ValidateRegionAreas(
        List<HashSet<(int X, int Y)>> regions, int nationChunkCount,
        GeographicConfig config)
    {
        var minArea = (int)(nationChunkCount * config.Region.MinAreaPct);
        var result = new List<HashSet<(int X, int Y)>>(regions);
        var changed = true;
        var maxIterations = 10;
        while (changed && maxIterations > 0)
        {
            changed = false;
            maxIterations -= 1;
            var i = 0;
            while (i < result.Count)
            {
                if (result[i].Count < minArea && result.Count > 1)
                {
                    var mergeTarget = FindAdjacentRegion(result[i], result, i);
                    if (mergeTarget >= 0)
                    {
                        result[mergeTarget].UnionWith(result[i]);
                        result.RemoveAt(i);
                        changed = true;
                        continue;   // same i — shifted element next
                    }
                }
                i += 1;
            }
        }
        return result;
    }

    public static (int MinX, int MinY, int MaxX, int MaxY) ComputeBounds(
        HashSet<(int X, int Y)> chunks)
    {
        if (chunks.Count == 0) return (0, 0, 0, 0);
        return (chunks.Min(c => c.X), chunks.Min(c => c.Y),
                chunks.Max(c => c.X), chunks.Max(c => c.Y));
    }

    /// <summary>generate_regions — global dense region ids across nations in
    /// ascending nation order; region_map filled sorted (patched Python).</summary>
    public static (Dictionary<(int X, int Y), int> RegionMap,
                   List<(int X, int Y)> RegionMapOrder,
                   Dictionary<int, RegionData> Metadata)
        GenerateRegions(NationGenerator.OrderedNationMap nationMap,
                        Dictionary<int, NationData> nationMetadata,
                        long seed, GeographicConfig config)
    {
        var nationTerritories = new Dictionary<int, HashSet<(int, int)>>();
        foreach (var pos in nationMap.Order)
        {
            var nid = nationMap.Map[pos];
            if (!nationTerritories.TryGetValue(nid, out var set))
                nationTerritories[nid] = set = new HashSet<(int, int)>();
            set.Add(pos);
        }

        var regionMap = new Dictionary<(int, int), int>();
        var regionMapOrder = new List<(int X, int Y)>();
        var regionMetadata = new Dictionary<int, RegionData>();
        var nextRegionId = 0;

        foreach (var nationId in nationTerritories.Keys.OrderBy(k => k))
        {
            var territory = nationTerritories[nationId];
            var nationData = nationMetadata.GetValueOrDefault(nationId);
            if (nationData is null || territory.Count == 0)
                continue;

            var regionSeed = seed + nationId * 1000 + 100000;
            var regionCount = DetermineRegionCount(nationId, territory.Count,
                                                   regionSeed, config);
            var rawRegions = GeoNoise.VoronoiSubdivide(
                territory, regionCount, regionSeed,
                noiseAmplitude: config.Region.DeformAmplitude,
                noiseFrequency: config.Region.DeformFrequency);
            var fixedRegions = FixContiguity(rawRegions);
            var validatedRegions = ValidateRegionAreas(fixedRegions, territory.Count, config);

            for (var i = 0; i < validatedRegions.Count; i++)
            {
                var regionChunks = validatedRegions[i];
                if (regionChunks.Count == 0) continue;   // consumes index i, no id

                var rid = nextRegionId;
                nextRegionId += 1;
                var identity = AssignIdentity(i, nationId, regionSeed);

                regionMetadata[rid] = new RegionData
                {
                    RegionId = rid,
                    Name = "",
                    NationId = nationId,
                    Identity = identity,
                    ChunkCount = regionChunks.Count,
                    Bounds = ComputeBounds(regionChunks),
                };

                nationData.RegionIds.Add(rid);

                foreach (var pos in regionChunks.OrderBy(p => p))   // sorted (patched)
                {
                    regionMap[pos] = rid;
                    regionMapOrder.Add(pos);
                }
            }
        }

        return (regionMap, regionMapOrder, regionMetadata);
    }
}
